Team9Project/src/com/indragie/cmput301as1/ElasticSearchAPIClient.java

```diff
@@ -0,0 +1,440 @@
+		 */
+		public void onFailure(Request request, Response response, IOException e);
+	}
+
+	/**
+	 * Interface for an object that can deserialize an ElasticSearch document
+	 * model from an HTTP response.
+	 * @param <T> The type of the document.
+	 */
+	private interface DocumentDeserializer<T extends ElasticSearchDocument> {
+		/**
+		 * Creates an ElasticSearch document from an HTTP response.
+		 * @param response The HTTP response.
+		 * @return An ElasticSearch document.
+		 */
+		T documentFromResponse(Response response);
+	}
+
+	//================================================================================
+	// Classes
+	//================================================================================
+	
+	/**
+	 * An exception that is thrown when an HTTP request to the ElasticSearch API fails.
+	 */
+	public class RequestFailedException extends Exception {
+		private static final long serialVersionUID = 1L;
+		
+		/**
+		 * The HTTP response that caused this exception.
+		 */
+		private Response response;
+		
+		/**
+		 * Creates a new instance of {@link RequestFailedException}
+		 * @param message The exception message.
+		 * @param response The HTTP response that caused this exception.
+		 */
+		public RequestFailedException(String message, Response response) {
+			super(message);
+			this.response = response;
+		}
+		
+		/**
+		 * @return The HTTP response that caused this exception.
+		 */
+		public Response getResponse() {
+			return response;
+		}
+	}
+	
+	/**
+	 * Wraps {@link com.squareup.okhttp.Call} with additional logic for deserializing
+	 * the document from JSON.
+	 * @param <T> The type of the document.
+	 */
+	public class APICall<T extends ElasticSearchDocument> {
+		//================================================================================
+		// Properties
+		//================================================================================
+		/**
+		 * The underlying {@link com.squareup.okhttp.Call} instance.
+		 */
+		private Call call;
+		
+		/**
+		 * Optional deserializer used to deserialize a document from JSON.
+		 */
+		private DocumentDeserializer<T> deserializer;
+		
+		//================================================================================
+		// Constructors
+		//================================================================================
+		/**
+		 * Creates a new instance of {@link APICall}
+		 * @param call The underlying {@link com.squareup.okhttp.Call} instance.
+		 * @param deserializer Optional deserializer used to deserialize a document from JSON.
+		 */
+		protected APICall(Call call, DocumentDeserializer<T> deserializer) {
+			this.call = call;
+			this.deserializer = deserializer;
+		}
+		
+		//================================================================================
+		// API
+		//================================================================================
+		/**
+		 * Cancels the request, if possible.
+		 */
+		public void cancel() {
+			call.cancel();
+		}
```

Comment:
Not at the moment, its more of a hypothetical use case if we did want to cancel an in-progress request. I'm not sure if that'll come up anywhere in this app though.